File: src/day6.rs
use crate::util::read_file;
use itertools::Itertools;
use std::collections::HashMap;

pub fn run() {
    println!("---- DAY 6 ----");
    part_1();
    part_2();
}

fn part_1() {
    let lines = read_file("inputs/day6.txt");
    let mut group = String::new();
    let mut total: i32 = 0;
    for line in lines.split('\n') {
        if line == "" {
            total += group.chars().unique().count() as i32;
            group.clear();
        }
        group.push_str(format!("{}", line).as_str());
    }
    println!("The sum is [{}]", total);
}

fn part_2() {
    let lines = read_file("inputs/day6.txt");
    let mut group = String::new();
    let mut total: i32 = 0;
    let mut people: i32 = 0;
    for line in lines.split('\n') {
        if line == "" {
            let mut cmap: HashMap<char, i32> = HashMap::new();
            for char in group.chars() {
                let mut val = match cmap.get(&char) {
                    Some(val) => *val,
                    None => 0i32
                };
                cmap.insert(char, val + 1);
            }
            total += cmap.keys().filter(|c| *cmap.get(&c).unwrap() == people).count() as i32;
            people = 0;
            cmap.clear();
            group.clear();
        } else {
            group.push_str(format!("{}", line).as_str());
            people += 1;
        }
    }
    println!("The sum is [{}]", total);
}

File: lib/day6.ex
defmodule Day6 do
    def run do
        groups = Input.read_file("day6")
            |> Enum.chunk_by(&(&1 == ""))
            |> Enum.filter(&(&1 != [""]))
            |> Enum.map(&(Enum.join(&1, " ")))
        part1(groups)
        part2(groups)
    end

    defp parse_groups([], total), do: total
    defp parse_groups([group|rest], total) do
        total = total + (group
            |> String.replace(" ", "")
            |> String.split("")
            |> Enum.filter(&(&1 != ""))
            |> Enum.uniq
            |> Enum.count)
        parse_groups(rest, total)
    end

    defp part1(groups) do
        IO.puts "6-1: The total is [#{parse_groups(groups, 0)}]"
    end

    defp parse_people(p, map \\ %{})
    defp parse_people([], map), do: map
    defp parse_people([person|people], map) do
        map = person |> String.split("")
            |> Enum.filter(&(&1 != ""))
            |> Enum.reduce(map, &(Map.update(&2, &1, 1, fn(v) -> v + 1 end)))
        parse_people(people, map)
    end

    defp parse_group(g, total \\ 0)
    defp parse_group([], total), do: total
    defp parse_group([group|rest], total) do
        people = String.split(group, " ")
        unique = parse_people(people)
        total = total + (unique
            |> Map.keys
            |> Enum.count(&(Map.get(unique, &1) == Enum.count(people)))
        )
        parse_group(rest, total)
    end

    defp part2(groups) do
        IO.puts "6-2: The total is [#{parse_group(groups)}]"
    end
end